### PR TITLE
Add keyword parameter to prevent interpreting uint16 arrays as uint8.

### PR DESCRIPTION
**Pull request by LandingEllipse**
Response to issue #352, better late than never...

When set to False, the new keyword parameter `interpret_uint8` accepted by `imwrite` can be set to false to inhibit the default write behaviour of PNGs whereby np.uint16 image arrays are cast to np.uint8 if their elements are compatible with the latter (i.e. all values fall within the range [0, 255]).

I have perhaps not found the most elegant way to implement this change, but it gets the job done. The indirection provided by the format writer inheritance hierarchy means that propagation of the new keyword parameter to `ndarray_to_pil` needs to go via the parent class `PillowFormat.Writer`. Depending on your approach to such things it might have been nice to be able to specify the default behaviour at the level of `PNGFormat`, however this would be complicated by the fact that `ndarray_to_pil` is a general function used indirectly by all children of `PillowFormat.Writer` and thus requires a default behaviour to be defined at that level or by the function itself. I opted for the latter, with the choice for `PillowFormat.Writer` to only present it with an override if the parameter has been passed in as a keyword argument from `imwrite`.

Basically I've tried to disturb as little of the codebase as possible; please let me know if you would prefer a different approach and I'll be happy to accommodate.

### CODE REVIEW

File: imageio/plugins/pillow.py [hunk 1]
```diff
@@ -286,6 +292,11 @@ class PNGFormat(PillowFormat):
         bits. In this case, given as a number between 1-256.
     dictionary (experimental): dict
         Set the ZLIB encoder dictionary.
+    interpret_uint8: bool
```
Comment (almarklein) [inline imageio/plugins/pillow.py]: I think renaming this to `prefer_uint8` would more accurately describe the behavior. What do you think?

File: imageio/plugins/pillow.py [hunk 2]
```diff
@@ -211,7 +211,13 @@ def _append_data(self, im, meta):
                 im = im[:, :, 0]
             self._written = True
             self._meta.update(meta)
-            img = ndarray_to_pil(im, self.format.plugin_id)
+            if "interpret_uint8" in self._meta:
+                img = ndarray_to_pil(
+                    im, self.format.plugin_id, self._meta["interpret_uint8"]
```
Comment (almarklein) [inline imageio/plugins/pillow.py]: Maybe use `self._meta.get("interpret_uint8", True)` instead of the `if-else`?

File: tests/test_pillow.py [hunk 3]
```diff
@@ -148,6 +148,12 @@ def test_png():
     im2 = imageio.imread(fnamebase + "1.png")
     assert im2.dtype == np.uint16
 
+    # issue #352 - prevent low-luma uint16 truncation to uint8
+    arr = np.full((32, 32), 255, dtype=np.uint16)  # values within range of uint8
+    imageio.imwrite(fnamebase + ".png", arr, interpret_uint8=False)
+    im = imageio.imread(fnamebase + ".png")
+    assert im.dtype == np.uint16
```
Comment (almarklein) [inline tests/test_pillow.py]: Could you please perform the three possible cases here: arg not given, arg set to `False`, arg set to `True`?

**Comment by almarklein** [comment]
Thanks! I made a few small comments.

File: imageio/plugins/pillow.py (hunk 1 above)
Comment (LandingEllipse) [reply imageio/plugins/pillow.py]: Perfect, the inaccuracy of "interpret" has been bugging me but I was stuck trying to come up with a better term.

File: imageio/plugins/pillow.py (hunk 2 above)
Comment (LandingEllipse) [reply imageio/plugins/pillow.py]: Aye, I agree this would be cleaner. Only consequence as noted in the PR is that this shifts specification of default behaviour to `PillowFormat.Writer._append_data`. To be fair, upon further consideration this doesn't really present an issue since the plugin's code base is relatively small; I'll implement the change.

File: tests/test_pillow.py (hunk 3 above)
Comment (LandingEllipse) [reply tests/test_pillow.py]: Will do!

**Comment by LandingEllipse** [comment]
Alright, all comments should have been responded to through commits; let me know if you have any further feedback.

**Comment by almarklein** [comment]
Nice, looking good! Thanks!
